libgloss/arc/uart-8250.c

```diff
@@ -0,0 +1,324 @@
+/* Transmit Holding Register Empty.  */
+#define	LSR_THRE	(1 << 5)
+/* Break Interrupt.  */
+#define	LSR_BI		(1 << 4)
+/* Framing Error.  */
+#define	LSR_FE		(1 << 3)
+/* Parity Error.  */
+#define	LSR_PE		(1 << 2)
+/* Overrun Error.  */
+#define	LSR_OE		(1 << 1)
+/* Data Ready.  */
+#define	LSR_DR		(1 << 0)
+
+/* Modem Status Register.  RO.  */
+#define	MSR		0x18
+/* Data Carrier Detect.  */
+#define	MSR_DCD		(1 << 7)
+/* Ring Indicator.  */
+#define	MSR_RI		(1 << 6)
+/* Data Set Ready.  */
+#define	MSR_DSR		(1 << 5)
+/* Clear To Send.  */
+#define	MSR_CTS		(1 << 4)
+/* Delta Data Carrier Detect.  */
+#define	MSR_DDCD	(1 << 3)
+/* Trailing Edge of Ring Indicator.  */
+#define	MSR_TERI	(1 << 2)
+/* Delta Data Set Ready.  */
+#define	MSR_DDSR	(1 << 1)
+/* Delta Clear To Send.  */
+#define	MSR_DCTS	(1 << 0)
+
+/* Main UART control structure.  */
+struct _uart_8250 {
+  /* Start of UART registers.  */
+  volatile char *base;
+  /* UART clock.  */
+  uint32_t clk;
+  /* Baud rate.  */
+  uint32_t baud;
+  /* If UART registers are mapped to AUX or memory.  */
+  int aux_mapped;
+  /* If UART is ready to use or not.  */
+  int ready;
+};
+
+struct _uart_8250 _uart_8250;
+
+/* Write 32-bit value to the UART register.  */
+static inline void
+_uart_8250_write_reg (const struct _uart_8250 *uart, uint32_t reg,
+		      uint32_t value)
+{
+  if (uart->aux_mapped)
+    __builtin_arc_sr (value, (uint32_t) uart->base + reg);
+  else
+    *(volatile uint32_t *)(uart->base + reg) = value;
+}
+
+/* Read 32-bit value from the UART register.  */
+static inline uint32_t
+_uart_8250_read_reg (const struct _uart_8250 *uart, uint32_t reg)
+{
+  if (uart->aux_mapped)
+    return __builtin_arc_lr ((uint32_t) uart->base + reg);
+  else
+    return *(volatile uint32_t *)(uart->base + reg);
+}
+
+/* Wait until all flags are set.  */
+static inline void
+_uart_8250_wait (const struct _uart_8250 *uart, uint32_t reg, uint32_t flags)
+{
+  while (1)
+    {
+      if ((_uart_8250_read_reg (uart, reg) & flags) == flags)
+	break;
+    }
+}
+
+/* Get one character from UART.  CR is converted to NL.  */
+static int
+_uart_8250_getc (const struct _uart_8250 *uart)
+{
+  char c;
+
+  _uart_8250_wait (uart, LSR, LSR_DR);
+
+  c = _uart_8250_read_reg (uart, RBR);
+
+  if (c == '\r')
+    c = '\n';
+
+  return c;
+}
+
+/* Put one character to UART.  CR is placed before NL.  */
+static void
+_uart_8250_putc (const struct _uart_8250 *uart, char c)
+{
+  if (c == '\n')
+    _uart_8250_putc (uart, '\r');
+
+  _uart_8250_wait (uart, LSR, LSR_TEMT | LSR_THRE);
+
+  _uart_8250_write_reg (uart, THR, c);
+}
+
+/*
+ * Setup UART control structure and following parameters:
+ *   - baudrate if clock and baudrate are passed to function.
+ *   - 8n1 (8 data bits, no parity bit, one stop bit).
+ *   - disable interrupts.
+ *   - disable FIFO.
+ *   - set Request To Send and Data Terminal Ready.
+ */
+int
+_uart_8250_setup (void *base, int aux_mapped, uint32_t clk, uint32_t baud)
```

Comment:
I guess `aux_mapped` parameter warrants some explanation as:
1. It has nothing to do with generic UART driver
2. Rare users are familiar with peculiarity of peripherals connection in case of at least some DW ARC subsystems.

Reply:
I added explanation about parameters to the function description.